model/riscv_insts_zfh.sail

```diff
@@ -0,0 +1,934 @@
+   & (mant [9] == bitone))
+}
+
+/* Either QNaN or SNan */
+val      f_is_NaN_H : bits(16) -> bool
+function f_is_NaN_H   xf16 = {
+  let (sign, exp, mant) = fsplit_H (xf16);
+  (  (exp == ones())
+   & (mant != zeros()))
+}
+
+val      fle_H : (bits(16), bits (16), bool) -> (bool, bits(5))
+function fle_H   (v1,       v2,        is_quiet) = {
+  let (s1, e1, m1) = fsplit_H (v1);
+  let (s2, e2, m2) = fsplit_H (v2);
+
+  let v1Is0    = f_is_neg_zero_H(v1) | f_is_pos_zero_H(v1);
+  let v2Is0    = f_is_neg_zero_H(v2) | f_is_pos_zero_H(v2);
+
+  let result : bool =
+    if (s1 == 0b0) & (s2 == 0b0) then
+      if   (e1 == e2)
+      then unsigned (m1) <=  unsigned (m2)
+      else unsigned (e1)  <  unsigned (e2)
+    else if (s1 == 0b0) & (s2 == 0b1)
+    then (v1Is0 & v2Is0)                         /* Equal in this case (+0=-0) */
+    else if (s1 == 0b1) & (s2 == 0b0)
+    then true
+    else
+      if   (e1 == e2)
+      then unsigned (m1) >=  unsigned (m2)
+      else unsigned (e1)  >  unsigned (e2);
+
+  let fflags = if is_quiet then
+                 if   (f_is_SNaN_H(v1) | f_is_SNaN_H(v2))
+                 then nvFlag()
+                 else zeros()
+               else
+                 if   (f_is_NaN_H(v1) | f_is_NaN_H(v2))
+                 then nvFlag()
+                 else zeros();
+
+  (result, fflags)
+}
+
+/* ****************************************************************** */
+/* Floating-point loads                                               */
+/* These are defined in: riscv_insts_fext.sail                        */
+
+/* ****************************************************************** */
+/* Floating-point stores                                              */
+/* These are defined in: riscv_insts_fext.sail                        */
+
+/* Binary ops with rounding mode */
+
+/* AST */
+
+union clause ast = F_BIN_RM_TYPE_H : (regidx, regidx, rounding_mode, regidx, f_bin_rm_op_H)
+
+/* AST <-> Binary encoding ================================ */
+
+mapping clause encdec =
+    F_BIN_RM_TYPE_H(rs2, rs1, rm, rd, FADD_H)                             if haveZfh()
+<-> 0b000_0010 @ rs2 @ rs1 @ encdec_rounding_mode (rm) @ rd @ 0b101_0011  if haveZfh()
+
+mapping clause encdec =
+    F_BIN_RM_TYPE_H(rs2, rs1, rm, rd, FSUB_H)                             if haveZfh()
+<-> 0b000_0110 @ rs2 @ rs1 @ encdec_rounding_mode (rm) @ rd @ 0b101_0011  if haveZfh()
+
+mapping clause encdec =
+    F_BIN_RM_TYPE_H(rs2, rs1, rm, rd, FMUL_H)                             if haveZfh()
+<-> 0b000_1010 @ rs2 @ rs1 @ encdec_rounding_mode (rm) @ rd @ 0b101_0011  if haveZfh()
+
+mapping clause encdec =
+    F_BIN_RM_TYPE_H(rs2, rs1, rm, rd, FDIV_H)                             if haveZfh()
+<-> 0b000_1110 @ rs2 @ rs1 @ encdec_rounding_mode (rm) @ rd @ 0b101_0011  if haveZfh()
+
+/* Execution semantics ================================ */
+
+function clause execute (F_BIN_RM_TYPE_H(rs2, rs1, rm, rd, op)) = {
+  let rs1_val_16b = nan_unbox_H (F(rs1));
```

Comment:
These will all need to use (a new) F_or_X_H but I guess it's fine to express it this way for now